### DISCUSSION TITLE
How to return a retry object?

### DISCUSSION THREAD

**Discussion by ArneMayer** [Xuanwo/backon#152] [Q&A]
Hi,

Inside a function, I am trying to execute and retry a task that I define by a closure.
Now, I would like to return the Retry object to the caller of the function, in order to give
out some kind of handle, that the caller can use to observe the retrying (for example by using notify or when).

I am having some trouble getting this to work.
I am getting this message:

`the size for values of type (dyn futures::Future<Output = Result<Any, ApplicationError>> + 'static) cannot be known at compilation time the trait Sized is not implemented for (dyn futures::Future<Output = Result<Any, ApplicationError>> + 'static)`

The code looks somewhat like this:

```
type RetryHandle<T, E> =
        Retry<
            backon::ExponentialBackoff,
            T,
            E,
            dyn Future<Output = Result<T, E>>,
            dyn FnMut() -> dyn Future<Output = Result<T, E>>,
        >,

async fn create_task() -> RetryHandle<Any, ApplicationError> {
    || async {
        // Do stuff..
    }.retry(backoff)
}
```

What am I doing wrong and how would one solve this?

**Comment by Xuanwo** [comment]
Hi, thanks for using `backon` first.

It's required by rustc that type must be `Sized`. We can't declare a type that contains `dyn Future<Output = Result<T, E>>` which is `!Sized`. The type of `Retry<...>` can't be specified at this time.

Would you like to try passing a function in `create_task()` instead?

**Reply by ArneMayer** [reply]
Hey, thank you for answering! What would that look like / What do you mean?

**Reply by Xuanwo** [reply]
Something like:

```
async fn create_task(f: fn(abc)) -> RetryHandle<Any, ApplicationError> {
    || async {
        // Do stuff..
    }.retry(backoff).notify(f)
}
```

**Reply by ArneMayer** [reply]
Good idea, I'll try it out.
